### ISSUE TITLE
[Fastspeech2 dataset] bucket_by_sequence_length

### ISSUE THREAD

**Issue by ronggong** [TensorSpeech/TensorFlowTTS#527]
Hi,

I think the fixed batch size with padded_batch of the dataset is not an efficient way to use the GPU capacity, because some batches will contain sequences with very unequal lengths, and the shorter ones need to be padded with long silence. I am working on using bucket_by_sequence_length to bucket batch where each one contains seqs with similar length. But the thing is the batch size will be variant from batch to batch. I wonder if the current loss calculation is done with the presumption of the fixed batch size? or is there any other pitfalls with variable batch size?

https://www.tensorflow.org/api_docs/python/tf/data/experimental/bucket_by_sequence_length

Maybe a relevant issue https://github.com/TensorSpeech/TensorFlowTTS/pull/455

**Comment by ronggong** [comment]
https://github.com/TensorSpeech/TensorFlowTTS/blob/31e446ab5ff4c1f3d228e424ce76acd113b676d0/tensorflow_tts/trainers/base_trainer.py#L804

That's the first problem I found if using a variable batch_size.

**Comment by dathudeptrai** [comment]
@ronggong so you want to use larger batch_size for a smaller sequence length ?. I think you can use fixed batch_size when using `bucket_by_sequence_length` to simplifying the problem :D.

**Comment by ronggong** [comment]
Yes, actually I am doing that. But I found using large batch for libritts converges badly. I wonder why...

**Comment by dathudeptrai** [comment]
> Yes, actually I am doing that. But I found using large batch for libritts converges badly. I wonder why...

large batch_size increases overfitting :D (in case training data is noise.)

**Comment by ronggong** [comment]
This is actually interesting effect. Because I see the 2d or 3d loss is calculated by reduce_mean of all dimensions except the batch, and then divided by the batch_size * n_gpu. So I assume the loss should be more stable when the batch_size increased, because large batch has more examples, so it should lead to a loss with less variance. 

I don't understand why this could lead to overfitting, and especially when the data is noisy...
